### ISSUE TITLE
GDExtension error: Condition "_instance_bindings != nullptr" is true.

### ISSUE THREAD

**Issue by Karmavil** [godotengine/godot#59117]
### Godot version

4.0.dev (0ce3fff)

### System information

Debian GNU/Linux bookworm/sid 64-bit, NVIDIA GM206 [GeForce GTX 960] (driver v470.103.01)

### Issue description

Hi. 
Using godot alpha I'm getting the following error when attempting to use get_node in ***_ready***:
```
ERROR: Condition "_instance_bindings != nullptr" is true.
```

I'm probably making use of _ready incorrectly but I am not sure.

Actually, what happens is that whatever code is inside *_ready* will be executed when the custom node is added to in the editor, not just when the game is launched. 
Is that the desired behavior for in example initialize default values for "exported" variables?

What this implies is that errors will be found when the node is added if the node needs some child and there are no find conditionals about them.
I added a project example I was creating to report a possible error I found in 3.x that might be happening in 4.0 and I faced this error. 
Am I doing something wrong here?

### Steps to reproduce

* Create a gdextension
* Add the custom node to the scene
* Define virtual _ready override
* Use `get_node` in the *_ready* method

### Minimal reproduction project

git clone --recursive -b _instance_bindings_nullptr https://github.com/Karmavil/minimal.git

**Comment by Karmavil** [comment]
Oh I am not binging my **_timer** property.. I hope that fix it

**Comment by voylin** [comment]
> Oh I am not binding my **_timer** property.. I hope that fix it

Did this fix the issue?

**Comment by Karmavil** [comment]
No yet.. I cannot find the way to pass a node reference as argument.
However, the property is private and it is only called by this class. Should I bind it?

I tried defining the set/get methods and bind them instead but as I said I have troubles passing a Node * as parameter. I also tried using a Ref but it doesn't work. 

TL;DR
I saw different comments about the use of Ref and the godot-cpp master example actually define a class that inherit RefCounted but the point is that I don't pretend to pass my custom node as argument I just want to pass a Node as argument to a function, but yeap I tried using Ref<godot:Timer> as well as Ref<godot::Node> and the error are not just more difficult to understand, a variable of that type is really hard to work with.

**Comment by Karmavil** [comment]
I defined the get/set using node pointers not `Ref`s methods and this is what I tried to bind:
```
    ClassDB::bind_method(D_METHOD("getTimerNode"), &minimal::Bug::getTimerNode);
    ClassDB::bind_method(D_METHOD("setTimerNode", "timer"), &minimal::Bug::setTimerNode);
    ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "_timer"), "setTimerNode", "getTimerNode");
```

But I get all this errors from Variant conversions:
`error: conversion from ‘const godot::Variant’ to ‘godot::Timer*’ is ambiguous`
`... operator int64_t() const;  ... `
`... operator int32_t() const; ...`
`... operator uint32_t() const; ...`
`... operator Object *() const; ...`

**Comment by Karmavil** [comment]
And if I bind this way:
```
    using namespace godot;
    ClassDB::bind_method(D_METHOD("nextSlide"), &minimal::Bug::nextSlide);
    ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "_timer"), "setTimerNode", "getTimerNode");
```
I cannot find the set method
```

ERROR: Setter method not found for property.
   at: add_property (src/core/class_db.cpp:78)
ERROR: Condition "_instance_bindings != nullptr" is true.
   at: set_instance_binding (core/object/object.cpp:1744)
```

**Comment by Karmavil** [comment]
@voylin The error is in the way I am binding the signal from a child to a function in my class.
I am not very familiar with this way to bind so if you wind the way let me know, as soon as I fix it I will close this issue.
Documentation is welcomed

**Comment by Karmavil** [comment]
Well bad news. Connecting the signal properly didn't fix the _instance_binding error

**Comment by voylin** [comment]
@Karmavil Can you update the minimal reproduction project, that way I can take a look at what you have now.

**Comment by Karmavil** [comment]
it is updated, but the error is still there. Use the branch _instance_bindings:
`git clone --recursive -b _instance_bindings_nullptr https://github.com/Karmavil/minimal.git`

I am building a godot without tools to verify this error is only shown in the editor

**Comment by voylin** [comment]
@Karmavil I can run get_node without any problems, are you running the alpha 4 build?

**Comment by Karmavil** [comment]
Well yes and no.. I'm getting a child node and connecting its timeout signal. That works fine but there is an error displayed.
That error is also shown with a build without tools, it is shown in the console not a big deal

**Comment by Karmavil** [comment]
I guess the most important aspect of this issue to be answered is if *_ready* should be triggered by the editor or it should be called only when the game is running, because that makes a lot of errors.

In example, if I remove the conditionals that I put in *_ready* when I add this node for first time (without childrens) it triggers an error because _ready is executed and there is no childrens
```
 
ERROR: Node not found: "Timer" (relative to "/root/@@15860/@@656/@@657/@@666/@@668/@@674/@@680/@@681/@@682/@@704/@@705/@@745/@@746/@@8870/@@8696/@@8697/@@8698/@@8699/@@8700/@@8701/Main/Bug").
```
@voylin do you suggest to update the Issue title? Because the error is still happening and I am not sure if it is causing this

**Comment by voylin** [comment]
@Karmavil Not sure about the title, but as I can't replicate the issue, my advice would be to put that node in a `onready var`.

**Comment by Karmavil** [comment]
Thank you, I will consider it

**Comment by Karmavil** [comment]
@voylin thank you very much for participate, in this triage (I guess that is) I have a question.
I will update the code adding one of this custom modules with a Timer node on it, and another without children, I will really appreciate feedback on that because I get a clear error here. Thank you in advance
Just give me a second to push the code

**Comment by Karmavil** [comment]
The error is gone. It was some commit at godot between 
this one: cfd21adf64087c750dde89191bcfcf4d166adc85
and this one: 7cabf49e67fbdcacf3af2db7361a70ea0ee766f6

Or godot-cp between:
this: b869ef660f585339dad23d502e70fcad5db78406
and this: 13603e40509800da02fbc7df29eff19e91b29158

If you're still having this error I will recommend to update godot-cpp

About the _action method being executed every time the tree is updated in the editor I will recommend the use of gdscript for that specific method because I still don't have an answer if it is a bug or a expected behavior. Or open a new Issue because this one got very extensive and out of topic

**Comment by nathanfranke** [comment]
I can still reproduce this in https://github.com/godotengine/godot/issues/62152 (MRP is there).
From what I can see this error is harmless, but there's probably something going wrong.

**Comment by Karmavil** [comment]
Thank you @nathanfranke for report an update on this issue.
I cannot take a look at it right now, my apologies

**Comment by nathanfranke** [comment]
I can actually reproduce in https://github.com/godotengine/godot-cpp/tree/master/test

```
$ godot4 --path test/demo
Godot Engine v4.0.alpha12.official.2c11e6d9e - https://godotengine.org
Vulkan API 1.2.0 - Using Vulkan Device #0: AMD - AMD Radeon RX 6800 XT
 
Constructor.
Signal bind

Static method calls
  static (109) 109
  void static
Instance method calls
  Simple func called.
  Simple const func called.
  Return something called.
  returned some string
  Return something const called.
```
```
ERROR: Condition "_instance_bindings != nullptr" is true.
   at: set_instance_binding (core/object/object.cpp:1665)
```
```
  returned const root:[Window:23941087638]
ExampleRef created.
  returned ref [ExampleRef:-9223372012225822184]
VarArg method calls
ExampleRef created.
ExampleRef created.
  Example ref checks called with value: -9223372012209045117, returning value: -9223372012192267910
  sending ref:  -9223372012209045117 returned ref:  -9223372012192267910
  Varargs (Variant return) called with 4 arguments
  vararg args 4
  Varargs (int return) called with 4 arguments
  vararg_nv ret 42
  Varargs (no return) called with 4 arguments
Method calls with default values
  defval (300) 300
  defval (250) 250
  defval (150) 150
Array and Dictionary
ExampleRef destroyed.
  test array [1, 2]
  test dictionary {"hello":"world", "foo":"bar"}
Properties
  custom position is (0, 0)
  custom position now is (50, 50)
Constnts
  FIRST 0
  ANSWER_TO_EVERYTHING 42
  CONSTANT_WITHOUT_ENUM 314
ExampleRef destroyed.
```

**Comment by Faless** [comment]
This seems to be an issue in godot-cpp and not godot itself which is instead warning of the wrong godot-cpp behavior.

See https://github.com/godotengine/godot-cpp/issues/679 and https://github.com/godotengine/godot-cpp/pull/798 for a potential fix.

Closing as this is tracked in its own repo ( godotengine/godot-cpp#679 )
